tests/testthat/test-parallel.R

```diff
@@ -0,0 +1,381 @@
+  x4 <- c(a = 1)
+  expect_named(pmap(list(x4, 1:2), `+`, .parallel = TRUE), c("a", "a"))
+})
+
+test_that("avoid expensive [[ method on data frames", {
+  local_bindings(
+    `[[.mydf` = function(x, ...) stop("Not allowed!"),
+    .env = global_env()
+  )
+
+  df <- data.frame(x = 1:2, y = 2:1)
+  class(df) <- c("mydf", "data.frame")
+
+  expect_equal(pmap(df, list, .parallel = list(`[[.mydf` = `[[.mydf`)), list(list(x = 1, y = 2), list(x = 2, y = 1)))
+  expect_equal(pmap_lgl(df, ~ TRUE, .parallel = list(`[[.mydf` = `[[.mydf`)), c(TRUE, TRUE))
+  expect_equal(pmap_int(df, ~ 2, .parallel = list(`[[.mydf` = `[[.mydf`)), c(2, 2))
+  expect_equal(pmap_dbl(df, ~ 3.5, .parallel = list(`[[.mydf` = `[[.mydf`)), c(3.5, 3.5))
+  expect_equal(pmap_chr(df, ~ "x", .parallel = list(`[[.mydf` = `[[.mydf`)), c("x", "x"))
+})
+
+test_that("pmap works with empty lists", {
+  expect_identical(pmap(list(), ~ 1, .parallel = TRUE), list())
+})
+
+test_that("preserves S3 class of input vectors (#358)", {
+  date <- as.Date("2018-09-27")
+  expect_identical(pmap(list(date), identity, .parallel = TRUE), list(date))
+})
+
+test_that("works with vctrs records (#963)", {
+  x <- new_rcrd(list(x = c(1, 2), y = c("a", "b")))
+  out <- list(new_rcrd(list(x = 1, y = "a")), new_rcrd(list(x = 2, y = "b")))
+  expect_identical(pmap(list(x, 1, 1:2), ~ .x, .parallel = TRUE), out)
+})
+
+test_that("don't evaluate symbolic objects (#428)", {
+  pmap(list(exprs(1 + 2)), ~ testthat::expect_identical(.x, quote(1 + 2)), .parallel = TRUE)
+  pwalk(list(exprs(1 + 2)), ~ testthat::expect_identical(.x, quote(1 + 2)), .parallel = TRUE)
+  expect_true(TRUE) # so the test is not deemed empty and skipped
+})
+
+# imap ----------------------------------------------------------------------
+
+test_that("atomic vector imap works", {
+  x <- 1:3 %>% set_names()
+  expect_true(all(imap_lgl(x, `==`, .parallel = TRUE)))
+  expect_length(imap_chr(x, paste, .parallel = TRUE), 3)
+  expect_equal(imap_int(x, ~ .x + as.integer(.y), .parallel = TRUE), x * 2)
+  expect_equal(imap_dbl(x, ~ .x + as.numeric(.y), .parallel = TRUE), x * 2)
+  expect_equal(imap_vec(x, ~ .x + as.numeric(.y), .parallel = TRUE), x * 2)
+})
+
+# map_at --------------------------------------------------------------------
+
+test_that("map_at() works with tidyselect", {
+  skip_if_not_installed("tidyselect")
+  local_options(lifecycle_verbosity = "quiet")
+
+  x <- list(a = "b", b = "c", aa = "bb")
+  one <- map_at(x, vars(a), toupper, .parallel = TRUE)
+  expect_identical(one$a, "B")
+  expect_identical(one$aa, "bb")
+  two <- map_at(x, vars(tidyselect::contains("a")), toupper, .parallel = TRUE)
+  expect_identical(two$a, "B")
+  expect_identical(two$aa, "BB")
+})
+
+test_that("negative .at omits locations", {
+  x <- c(1, 2, 3)
+  out <- map_at(x, -1, ~ .x * 2, .parallel = TRUE)
+  expect_equal(out, list(1, 4, 6))
+})
+
+# ---------------------------------------------------------------------------
+
+mirai::daemons(0)
```

Comment:
This shouldn't be necessary, right? (Because of the `on.exit()` call above. Or is just just belt-and-braces for someone running the file line-by-line?

Reply:
I just want to be extra careful in case the tests are run in parallel etc. `daemons(0)` always returns zero - it's just a no-op if there's nothing set up.